lib/Sema/CSDiagnostics.h

```diff
@@ -465,15 +465,20 @@ class SuperclassRequirementFailure final : public RequirementFailure {
 /// Call to `foo` is going to be diagnosed as missing `q:`
 /// and having extraneous `a:` labels, with appropriate fix-its added.
 class LabelingFailure final : public FailureDiagnostic {
-  ArrayRef<Identifier> CorrectLabels;
+  ArgumentRelabeling Relabeling;
```

Comment:
I went through this a couple of times and I can say that I like previous approach better. I think we should record individual `RelabelArgument` fixes for each incorrect position so that content of `ArgumentRelabelingItem` could be inlined into the fix itself and then have a `RelabelArgument::coaleseAndDiagnose` method which would group labeling fixes if necessary e.g., if majority of labels are incorrect, and produce a single diagnostic. WDYT?

Reply:
I thought about that today.

Basically, I think it's good.

Certainly, it makes sense to generate fixes for each label that has an problem, and then aggregate them at the time of the `coalesceAndDaignose` where the fixes are translated into diagnostics.

As a constraint system, it is easier to understand the presence of fixes for each label when looking at the intermediate state with `-debug-constraints`, and the impact value of each fix is constant.

The implementation of `matchCallArguments` is also simplified around the following.

https://github.com/apple/swift/blob/e26763d56be595f2a19bf9022e890696def118f8/lib/Sema/CSSimplify.cpp#L688-L713

However, some changes in diagnostics are inevitable because other information about reordering fix, extraneous or missing arguments are not detectable in `RelabelArgument::coalesceAndDiagnose`.

(1). It can't control not aggregating label errors when there is a reorder fix.

For example, it would be:

```swift
func f(aa: Int, bb: Int, cc: Int, dd: Int, ee: Int) {}

f(aax: 0, bb: 1, dd: 3, cc: 2, eex: 5)
// incorrect argument labels in call (have 'aax:eex:', expected 'aa:ee:')
// fix-its: aax => aa, eex => ee

// argument 'cc:' must precede argument 'dd:'
```

but in my current implementation:

```swift
func f(aa: Int, bb: Int, cc: Int, dd: Int, ee: Int) {}

f(aax: 0, bb: 1, dd: 3, cc: 2, eex: 5)
// incorrect argument label in call (have 'aax:', expected 'aa:')
// fix-its: aax => aa

// incorrect argument label in call (have 'eex:', expected 'ee:')
// fix-its: eex => ee

// argument 'cc:' must precede argument 'dd:'
```

(2). It shows the given label and the expected label when it produces an aggregated label error. At this time, the given and expected labels cannot be reconstructed because the information of the extra argument and missing argument is not passed to the `RelabelArgument::coalesceAndDiagnose`. These are the cases represented as `None` in the `ArgumentRelabelingItem`. 

For a given label, It can get the AST information from the Locator. So the extra argument is fine. To begin with, currently it doesn't diagnose label issues when there is an extra argument, but I would like to do this at the same time in the future.

However, the label of the parameter corresponding to the missing argument can't be helped.

For example, it would be:

```swift
func f(aa: Int, bb: Int, cc: Int, dd: Int, ee: Int) {}

f(aa: 0, cc: 2, ddx: 3, eex: 3)
// incorrect argument labels in call (have 'ddx:eex:', expected 'dd:ee:')
// fix-its: ddx => dd, eex => ee

// missing argument for parameter 'bb' in call
```

but in my current implementation:

```swift
func f(aa: Int, bb: Int, cc: Int, dd: Int, ee: Int) {}

f(aa: 0, cc: 2, ddx: 3, eex: 3)
// incorrect argument labels in call (have 'aa:cc:ddx:eex:', expected 'aa:bb:cc:dd:ee:')
// fix-its: ddx => dd, eex => ee

// missing argument for parameter 'bb' in call
```

I can accept both cases.
It's rather simple and may be better.

With this policy, I can drop the `Optional` in the `ArgumentRelabelingItem`.

I will try to implement it.

Reply:
I think we can do it this way - aggregate only if there are more than two labeling failures and no other failures associated with arguments of the same call?

Also please note that "aggregate" diagnostic has to produce _all_ of the labels because it's about what call as a whole expects.

Reply:
If I let `RelabelArgument` have only contents of the `ArgumentRelabelingItem` as you said at start of this review thread, it can't produce an aggregated diagnosis that includes *all* the labels because `RelabelArgument` fix is not generated for the missing argument, extra argument or label where there is no problem.

I considered that and made one previous reply.

Am I misreading something?

Reply:
It would be easy enough to retrieve parameter labels from an overload associated with the call via `getOverloadChoice(getCalleeLocator(getLocator()))` (from FailureDiagnostic), maybe worth a separate fix/diagnostic even.

Reply:
I got it! Thanks for the advice.